contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,267 @@
+        address playerOne,
+        address playerTwo,
+        uint256 stake
+    );
+
+    event PlayerMove(
+        address indexed player,
+        bytes32 indexed token
+    );
+
+    event Draw(
+        bytes32 indexed token
+    );
+
+    event AllPlayersMoved(
+        bytes32 indexed token
+    );
+
+    event AllMovesRevealed(
+        bytes32 indexed token
+    );
+
+    event WinnerRewarded(
+        bytes32 indexed token,
+        address indexed winner,
+        uint256 winnings
+    );
+
+    event WithDraw(
+        address indexed withdrawer,
+        uint amount
+    );
+
+    event  PlayerWithdraws(
+        address indexed player,
+        bytes32 indexed token,
+        uint256 refund
+    );
+
+    uint8 constant ROCK = 1;
+    uint8 constant PAPER = 2;
+    uint8 constant SCISSORS = 3;
+
+    function deposit() external payable whenNotPaused {
+
+        balances[msg.sender] = balances[msg.sender].add(msg.value);
+
+        emit Deposit(msg.sender, msg.value);
+    }
+
+    function initialise(bytes32 gameToken, address playerOne, address playerTwo, uint stake) external whenNotPaused onlyOwner {
+
+        require(playerOne != msg.sender, "Caller cannot be player one");
+        require(playerTwo != msg.sender, "Caller cannot be player two");
+        require(playerOne != address(0), "Player one cannot be empty");
+        require(playerTwo != address(0), "Player two cannot be empty");
+        require(playerOne != playerTwo, "Player one and two are the same");
+
+        Game storage game = games[gameToken];
+
+        game.playerOne = playerOne;
+        game.playerTwo = playerTwo;
+        game.stake = stake;
+
+        emit GameInitialised(msg.sender, gameToken, playerOne, playerTwo, stake);
+    }
+
+    function play(bytes32 gameToken, bytes32 playerMoveHash) public whenNotPaused {
+
+        require(gameToken != bytes32(0), "Game token cannot be empty");
+        require(playerMoveHash != bytes32(0), "Move token cannot be empty");
+
+        Game storage game = games[gameToken];
+
+        require(game.stake != 0, GAME_NOT_FOUND_MSG);
+        require(msg.sender == game.playerOne || msg.sender == game.playerTwo, INVALID_PLAYER_MSG);
+
+        uint256 stakerBalance = balances[msg.sender];
+        require(stakerBalance >= game.stake, "Insufficient funds to play");
+
+        if (msg.sender == game.playerOne) {
+            require(game.playerOneMoveHash == bytes32(0), ALREADY_MOVED_MSG);
+            game.playerOneMoveHash = playerMoveHash;
+        }
+        if (msg.sender == game.playerTwo) {
+            require(game.playerTwoMoveHash == bytes32(0), ALREADY_MOVED_MSG);
+            game.playerTwoMoveHash = playerMoveHash;
+        }
+
+        emit PlayerMove(msg.sender, gameToken);
+
+        if(game.playerOneMoveHash != bytes32(0) && game.playerTwoMoveHash != bytes32(0)){
+            emit AllPlayersMoved(gameToken);
+        }
+
+        balances[msg.sender] = SafeMath.sub(stakerBalance, game.stake);
```

Comment:
I see. I was considering the SLOAD optimisation but I had a poor understanding in this context.

If I use balances[msg.sender] several times, it is an additional call. I assumed if game is already loaded into memory then I can call games's properties repeatedly without cost.

Reply:
You declared it as `storage`, it means that `game` is a storage pointer, not a copy of the whole struct in memory.
Given you use so few of the struct, copying it wholesale in memory would be inefficient.
But if `game` were in memory, it would still cost an `MLOAD` to read `game.stake`. Cheaper though.

Reply:
I see. Thank you. If it was loaded into memory then I understand that I can't write back to it.